rfcs/004 - Continuous Two-Way Synchronization RFC.md

```diff
@@ -0,0 +1,209 @@
+
+| Failure Mode | Cron | Event |
+| ------------ | ---- | ----- |
+| Stuck job >45min | The graphile worker library kills the job after 45 minutes and then the cron starts work again, which  may resolve correctly this time or it may be stuck again, but any new changes by the user to the DB will be considered. | The graphile worker library kills the job after 45 minutes and then no event for the next run is triggered, which causes this DB to be soft-locked until there is an engine redeploy. |
+| Engine dies | The cron keeps trying and failing to call the engine. The cron SQL function inserts start and end audit records immediately after each other, and does nothing. | Nothing happens at all. |
+| Database dies | Nothing happens at all. | The inner cron of the engine will keep trying to communicate to the DB and fail. This may or may not be handled correctly by our code. |
+| Engine deployed | The cron creates the graphile job and one of the two parallel versions of the engine takes the work (as this approach requires a SQL RPC call to start the engine) and the other doesn't work on it. If it is the old version that takes it, it may fail and no work will happen for 45 minutes, then it will continue. | The two engines both have their own inner cron unaware of each other and very likely they will both try to execute at the same time for 7-10 minutes. What they will do is undefined unless we thoroughly test it, but likely they will "fight" with each other undoing each other's work and potentially corrupting the DB. |
+
+Most of the failure modes when using the cron approach lead to delays in eventual consistency between the database and the cloud, while the event-based approach can lead to more catastrophic failures.
+
+Backing the cron approach would be a minor addition to the existing audit log to go with the graphile worker job creation. If we have the new cron logic and the rollback/commit calls add rows to the audit log when it starts and ends, they can check if the most recent of these records is a `start` one, implying that the engine is currently working on applying and synchronizing changes with the cloud, and would alter their behavior slightly. The `cron` would simply early exit in this case, as it implies there is an existing cron that has gone overtime, and so it should wait until the next execution of the cron to try again. `iasql_begin` would block until a paired `end` record exists and then continue with its work. `iasql_rollback` and `iasql_commit` should therefore never see this state after `iasql_begin` is called, but for defensiveness they should as well block until an `end` record exists and then do their own work.
```

Comment:
I need to think harder about this / better understand how `pg_cron` works, but to be able to represent autonomous transactions (https://github.com/iasql/iasql-engine/blob/main/CONTRIBUTING.md#how-iasql-postgresql-functions-work) the cron would need to support dblink or I guess that code lives in SQL anyway and there is no additional work here?

Reply:
So there would be a minor difference between the `iasql_commit()` function and what the cron calls, but they would be almost identical and I think it could simply be a "pre-amble" in the pg_cron cron function to just *not* call `iasql_commit()` if there's an `start_commit` without a paired `end_commit` within the last 45 minutes.

That starts to get into deeper implementation details, but I believe it will only be a couple of extra lines of pure SQL code (beyond the overhead of the wrapper function to pass to `pg_cron`).